src/Grpc.Net.Client/Balancer/Internal/StreamWrapper.cs

```diff
@@ -78,11 +80,30 @@ public override int WriteTimeout
 #endif
 
     public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken) =>
-        _inner.ReadAsync(buffer, offset, count, cancellationToken);
+        ReadAsync(buffer.AsMemory(offset, count), cancellationToken).AsTask();
 
 #if !NETSTANDARD2_0
-    public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default) =>
-        _inner.ReadAsync(buffer, cancellationToken);
+    public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
+    {
+        if (_initialSocketData != null && _initialSocketData.Count > 0)
```

Comment:
I didn't use a ReadOnlySequence because they're a pain to create. But you're right that it would reading easier.

I expect only a small amount of data from the SETTINGS frame. It's likely the list will have one array in it.